articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,234 @@
+---
+layout: default
+title: ROS 2 Launch XML Format
+permalink: articles/roslaunch_xml.html
+abstract:
+  The XML format for declarative launch descriptions in the ROS 2 launch system.
+author: '[Michel Hidalgo](https://github.com/hidmic)'
+published: false
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Authors: {{ page.author }}
+
+
+# ROS 2 Launch XML Format v0.1.0
+
+## Rationale
+
+As an alternative to a programmatic approach to the ROS 2 launch system's API, a declarative description features a WYSIWYG approach, easier to read, audit and maintain.
+This is the preferred approach for ROS 1 `roslaunch` launch files, thus some degree of familiarity is expected (and relied upon).
+
+The body of such a description is mainly comprised of statically declared launch actions with a prescribed configuration.
+To that, runtime value substitution is added in order to fullfill common dynamic (re)configuration needs like conditional execution, resource lookups, etc.
+It is intended for these entities to map to those of the underlying implementation, reducing support to file parsing.
+
+The choice of XML over other markup languages aims to ease the bridge between ROS and ROS 2 launch files.
+Widely available support for parsing XML in a myriad of languages and platforms also weighed (significantly) in the decision.
+
+## Static Description
+
+### Schema Definition
+
+```xml
+{% include_relative specs/launch.0.1.0.xsd %}
+```
+
+### Tags Semantics
+
+#### All Tags
+
+Every tag's execution can be conditioned on a boolean predicate via `if` or `unless` attributes.
+All action tags that can contain other action tags can scope ROS entities in a namespace via the `ns` attribute.
+
+#### `<launch>` Tag
+
+Root tag of any launch file.
+There must only be one `<launch>` tag on a given launch file.
+
+#### `<include>` Tag
```

Comment:
Can the include element be just a macro around xinclude that expands the Built-in Substitutions, or does it need to do more than that?

Reply:
So, as immediate needs, the `<include>` tag has to be able to do ROS namespacing and take arguments. That I'd think it's achievable by just injecting actions before and after included content. 

But on top of that, when I wrote the draft I thought that allowing inclusion of launch files written in different markups or even programmatically would provide a significant amount of flexibility (e.g. an XML launch including a Python launch).

Reply:
> That I'd think it's achievable by just injecting actions before and after included content.

What are `actions` in this context, is this an XML spec?

Reply:
Sorry, my bad for using launch terminology out of place. [Here](https://github.com/ros2/design/pull/163/files#diff-df7db447d3b09c67c5487a923f56a445R586) is the formal definition, but long story short actions are the way one describes the act of doing something in ROS2 launch and ultimately what makes up most of launch descriptions. In this context, `<include>` maps to an action, `<group>` to another and so on.

Reply:
@ruffsl I think it will be easier for you to contribute to this discussion if you skim over the architecture document for launch:

https://github.com/ros2/launch/blob/master/launch/doc/source/architecture.rst

It's not that long and it introduces most, if not all, of the concepts that the markup needs to capture, and some of the reasoning why they are capture as they are rather than leaning on language features (whether it be programming languages like Python or markup languages like XML) to accomplish certain tasks.